### PR TITLE
[WIP] Add support for type-casting parameters

### LINKED ISSUE #733: Add type-casts for Parameters
Since parameter values are going to be loaded from context (which is JSON-compatible), we should allow passing type-casting functions when creating a parameter.

Something like (arg name debatable):
```python
dt = Parameter('dt', type_cast_fn=lambda dt: pendulum.parse(dt))
```

We could also use output typing to do this automatically?

### PR DESCRIPTION
**Pull request by jlowin**
**Thanks for contributing to Prefect!**

Please describe your work and make sure your PR:

- [x] adds new tests (if appropriate)
- [x] updates `CHANGELOG.md` (if appropriate)
- [x] updates docstrings for any new functions or function arguments, including `docs/outline.toml` for API reference docs (if appropriate)

Note that your PR will not be reviewed unless all three boxes are checked.

## What does this PR change?
This PR adds support for type-casting Parameters, which are usually provided as strings/JSON. Closes #733.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -16,6 +16,7 @@ These changes are available in the [master branch](https://github.com/PrefectHQ/
 - Reduces the size of the `prefecthq/prefect` Docker image by ~400MB, which is now the base Docker image used in Flows - [#1648](https://github.com/PrefectHQ/prefect/pull/1648)
 - Add a new healthcheck for environment dependencies - [#1653](https://github.com/PrefectHQ/prefect/pull/1653)
 - Add default 30 second timeout to Client requests - [#1672](https://github.com/PrefectHQ/prefect/pull/1672)
+- Add type-casting for `Parameters` - [#1682](https://github.com/PrefectHQ/prefect/pull/1682)
```
Comment (cicdw) [inline CHANGELOG.md]: Would you mind cutting the 0.7.0 changelog and placing this in the "unreleased" section?

Comment (jlowin) [reply CHANGELOG.md]: Done in #1684

File: src/prefect/serialization/task.py [hunk 2]
```diff
@@ -129,5 +129,8 @@ class Meta:
     default = JSONCompatible(allow_none=True)
     required = fields.Boolean(allow_none=True)
     description = fields.String(allow_none=True)
+    cast = fields.Function(
```
Comment (cicdw) [inline src/prefect/serialization/task.py]: Because of this I'm going to hold off on merging until post 0.8.1

File: CHANGELOG.md [hunk 3]
```diff
@@ -10,6 +10,7 @@ These changes are available in the [master branch](https://github.com/PrefectHQ/
 
 ### Enhancements
 
+- Add type-casting for `Parameters` - [#1682](https://github.com/PrefectHQ/prefect/pull/1682)
```
Comment (cicdw) [inline CHANGELOG.md]: This should be moved

**Comment by jlowin** [comment]
This is actually still open because of the unresolved issues with handling type-cast results. I've experimented with using the new `ConstantResultHandler` but was waiting to see what we decide on a default storage backend.

**Comment by cicdw** [comment]
Ah OK gotcha, makes sense

**Comment by jlowin** [comment]
Closing this in favor of a potential path forward in #2760

**Pull request closed without merging**